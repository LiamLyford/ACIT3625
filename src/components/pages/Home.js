import React, { useEffect, useState } from 'react';
import { nanoid } from 'nanoid';

import Todo from '../Todo';
import Form from '../Form';
import FilterButton from '../FilterButton';
import DATA from '../../data.json';

const FILTER_MAP = {
    All: () => true,
    Active: task => !task.completed,
    Completed: task => task.completed
}
const FILTER_NAMES = Object.keys(FILTER_MAP);

function Home() {
    const [tasks, setTasks] = useState(DATA);
    const [filter, setFilter] = useState('All');
    const [warning, setWarning] = useState('');

    // if there are tasks stored for the user, then load those. Otherwise, start with the defaults
    useEffect(() => {
        const data = localStorage.getItem('listOfTasks');
        if (data) {
            setTasks(JSON.parse(data));
        }
    }, []);


    useEffect(() => {
        localStorage.setItem('listOfTasks', JSON.stringify(tasks));
    });

    function addTask(name) {
        if (name === '') {
            setWarning('You must enter a name')
        } else {
            var newTask = { id: "todo-" + nanoid(), name: name, completed: false };
            // var temp = tasks;
            // temp.push(newTask);
            // console.log(typeof(tasks))
            // console.log(typeof(temp)); 
            // setTasks(temp);
            // console.log(tasks);
            // console.log(typeof(tasks))
            setTasks([...tasks, newTask]);
            setWarning('');
        }
    }

    const toggleTaskCompleted = (id) => {
        const updatedTasks = tasks.map(task => {
            if (id === task.id) {
                return { ...task, completed: !task.completed }
            }
            return task
        });
        setTasks(updatedTasks)
    }

    const deleteTask = (id) => {
        const remainingTasks = tasks.filter(task => id !== task.id);
        setTasks(remainingTasks);
    }

    const editTask = (id, newName) => {
        const updatedTasks = tasks.map(task => {
            if (task.id === id) {
                return { ...task, name: newName }
            }
            return task;
        });
        setTasks(updatedTasks);
    }

    const taskList = tasks
        .filter(task => FILTER_MAP[filter](task))
        .map(task =>
            <Todo
                id={task.id}
                name={task.name}
                completed={task.completed}
                key={task.id}
                toggleTaskCompleted={toggleTaskCompleted}
                deleteTask={deleteTask}
                editTask={editTask}
            />);

    const tasksNoun = taskList.length !== 1 ? 'tasks' : 'task';
    const headingText = `${taskList.length} ${tasksNoun} remaining`;

    const filterList = FILTER_NAMES.map(name => (
        <FilterButton
            key={name}
            name={name}
            isPressed={name === filter}
            setFilter={setFilter}
        />
    ))

    return (
        <div className="todoapp stack-large">
            <h1>TodoMatic</h1>
            <Form addTask={addTask} />
            <div style={{ textAlign: 'center' }}><span className="name-warning">{warning}</span></div>
            <div className="filters btn-group stack-exception">
                {filterList}
            </div>
            <div style={{ display: "flex", justifyContent: 'space-between', alignItems: 'baseline' }}>
                <h2 id="list-heading">
                    {headingText}
                </h2>
                <button type="button" className="btn toggle-btn" onClick={() => { localStorage.clear(); setTasks([]) }}>
                    <span className="visually-hidden">Clear </span>
                    <span>Clear All Tasks</span>
                    <span className="visually-hidden"> tasks</span>
                </button>
            </div>
            <ul
                className="todo-list stack-large stack-exception"
                aria-labelledby="list-heading"
            >
                {taskList}
            </ul>
        </div>
    );
}

export default Home;
