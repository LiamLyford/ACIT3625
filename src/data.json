[
    { "id": "todo-0", "name": "Eat", "status": 0 },
    { "id": "todo-1", "name": "Sleep", "status": 1 },
    { "id": "todo-2", "name": "Repeat", "status": 2 }
]