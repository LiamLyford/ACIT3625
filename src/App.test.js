import React from 'react';
import renderer from 'react-test-renderer';
import Home from './components/pages/Home';
import data from './data.json';

describe('Home snapshot', () => {
      test('Home snapshot test', () => {
      const component = renderer.create(<Home />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    })
});

const itemNum = 0
describe('Data tests on item '+ itemNum + ' in array', () => {
    test('id contains todo', () => {
        expect(data[itemNum].id).toContain('todo');
    });
    test('object has id property', () => {
        expect(data[itemNum]).toHaveProperty('id');
    });
    test('object has name property', () => {
        expect(data[itemNum]).toHaveProperty('name');
    });
    test('object has completed property', () => {
        expect(data[itemNum]).toHaveProperty('completed');
    });
})



