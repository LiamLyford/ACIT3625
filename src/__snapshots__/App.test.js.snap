// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot Home snapshot test 1`] = `
<div
  className="todoapp stack-large"
>
  <h1>
    TodoMatic
  </h1>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="label-wrapper"
    >
      <label
        className="label__lg"
        htmlFor="new-todo-input"
      >
        What needs to be done?
      </label>
    </h2>
    <input
      autoComplete="off"
      className="input input__lg"
      id="new-todo-input"
      name="text"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="btn btn__primary btn__lg"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <span
      className="name-warning"
    >
      
    </span>
  </div>
  <div
    className="filters btn-group stack-exception"
  >
    <button
      aria-pressed={true}
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        All
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
    <button
      aria-pressed={false}
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        Active
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
    <button
      aria-pressed={false}
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        Completed
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
  </div>
  <div
    style={
      Object {
        "alignItems": "baseline",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <h2
      id="list-heading"
    >
      3 tasks remaining
    </h2>
    <button
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Clear 
      </span>
      <span>
        Clear All Tasks
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
  </div>
  <ul
    aria-labelledby="list-heading"
    className="todo-list stack-large stack-exception"
  >
    <li
      className="todo stack-small"
    >
      <div
        className="stack-small"
      >
        <div
          className="c-cb"
        >
          <input
            defaultChecked={true}
            id="todo-0"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="todo-label"
            htmlFor="todo-0"
          >
            Eat
          </label>
        </div>
        <div
          className="btn-group"
        >
          <button
            className="btn"
            onClick={[Function]}
            type="button"
          >
            Edit 
            <span
              className="visually-hidden"
            >
              Eat
            </span>
          </button>
          <button
            className="btn btn__danger"
            onClick={[Function]}
            type="button"
          >
            Delete 
            <span
              className="visually-hidden"
            >
              Eat
            </span>
          </button>
        </div>
      </div>
    </li>
    <li
      className="todo stack-small"
    >
      <div
        className="stack-small"
      >
        <div
          className="c-cb"
        >
          <input
            defaultChecked={false}
            id="todo-1"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="todo-label"
            htmlFor="todo-1"
          >
            Sleep
          </label>
        </div>
        <div
          className="btn-group"
        >
          <button
            className="btn"
            onClick={[Function]}
            type="button"
          >
            Edit 
            <span
              className="visually-hidden"
            >
              Sleep
            </span>
          </button>
          <button
            className="btn btn__danger"
            onClick={[Function]}
            type="button"
          >
            Delete 
            <span
              className="visually-hidden"
            >
              Sleep
            </span>
          </button>
        </div>
      </div>
    </li>
    <li
      className="todo stack-small"
    >
      <div
        className="stack-small"
      >
        <div
          className="c-cb"
        >
          <input
            defaultChecked={false}
            id="todo-2"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="todo-label"
            htmlFor="todo-2"
          >
            Repeat
          </label>
        </div>
        <div
          className="btn-group"
        >
          <button
            className="btn"
            onClick={[Function]}
            type="button"
          >
            Edit 
            <span
              className="visually-hidden"
            >
              Repeat
            </span>
          </button>
          <button
            className="btn btn__danger"
            onClick={[Function]}
            type="button"
          >
            Delete 
            <span
              className="visually-hidden"
            >
              Repeat
            </span>
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
